qiskit_ibm_runtime/noise_learner.py

```diff
@@ -0,0 +1,276 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Noise learner program."""
+
+from __future__ import annotations
+
+from dataclasses import asdict
+from typing import Any, Dict, Iterable, Optional, Union
+import logging
+
+from qiskit.circuit import QuantumCircuit
+from qiskit.providers import BackendV1, BackendV2
+from qiskit.primitives.containers import EstimatorPubLike
+from qiskit.primitives.containers.estimator_pub import EstimatorPub
+
+from .constants import DEFAULT_DECODERS
+from .runtime_job_v2 import RuntimeJobV2
+from .ibm_backend import IBMBackend
+from .options.estimator_options import EstimatorOptions
+from .options.noise_learner_options import NoiseLearnerOptions
+from .options.utils import remove_dict_unset_values, remove_empty_dict
+from .utils import validate_isa_circuits
+from .utils.utils import is_simulator
+
+from .fake_provider.local_service import QiskitRuntimeLocalService
+from .qiskit_runtime_service import QiskitRuntimeService
+from .provider_session import get_cm_session
+
+
+# pylint: disable=unused-import,cyclic-import
+from .session import Session
+from .batch import Batch
+
+logger = logging.getLogger(__name__)
+
+
+class NoiseLearner:
```

Comment:
I have reservations about how generic this name is, but unfortunately have nothing more constructive to say at this time. The concern is that we'll one day want a different thing to be learned, in which case we'll regret making this name not specific to Sparse Pauli learning.

Reply:
I don't think I fully like the idea of inheriting from `BasePrimitiveV2`, because that would make `NoiseLearner` be a primitive both as an object and conceptually (imagine a user that looks at it for the first time and tries to understand what it is). However, I agree that there is a lot of duplicate code between this implementation of noise learner and base primitives V2, which I like even less than inheriting. 
Therefore, I am okay with inheriting, but raising the point that we may want to think whether we have a missing concept in our codebase and are just shoving it into `BasePrimitiveV2`

Reply:
> we may want to think whether we have a missing concept in our codebase and are just shoving it into BasePrimitiveV2

Yeah we should have a generic `ProgramRunner` now that we can call more than just primitives. But that can be another PR.